File: src/main/java/net/maple/handlers/user/UserAbilityUpRequestHandler.java
package net.maple.handlers.user;

import client.Character;
import client.Client;
import client.player.StatType;
import net.maple.handlers.PacketHandler;
import util.packet.PacketReader;

import java.util.ArrayList;
import java.util.List;

public class UserAbilityUpRequestHandler extends PacketHandler {

    @Override
    public void handlePacket(PacketReader reader, Client c) {
        Character chr = c.getCharacter();
        reader.readInteger(); // ?

        int type = reader.readInteger();
        if (chr.getAp() > 0) {
            chr.decAP();
            List<StatType> statTypes = new ArrayList<>();
            statTypes.add(StatType.AP);

            if (type == StatType.STR.getStat()) {
                chr.incStrength();
                statTypes.add(StatType.STR);
            } else if (type == StatType.DEX.getStat()) {
                chr.incDexterity();
                statTypes.add(StatType.DEX);
            } else if (type == StatType.INT.getStat()) {
                chr.incIntelligence();
                statTypes.add(StatType.INT);
            } else if (type == StatType.LUK.getStat()) {
                chr.incLuck();
                statTypes.add(StatType.LUK);
            }

            chr.updateStats(statTypes, true);
        }
    }
}

File: src/main/java/net/maple/handlers/user/UserAbilityUpRequestHandler.kt
package net.maple.handlers.user

import client.Client
import client.player.StatType
import net.maple.handlers.PacketHandler
import util.packet.PacketReader
import java.util.*

class UserAbilityUpRequestHandler : PacketHandler {

    override fun handlePacket(reader: PacketReader, c: Client) {
        val chr = c.character

        reader.readInteger() // timestamp
        val type = reader.readInteger()

        if (chr.ap > 0) {
            chr.ap--

            when (type) {
                StatType.STR.stat -> chr.strength++
                StatType.DEX.stat -> chr.dexterity++
                StatType.INT.stat -> chr.intelligence++
                StatType.LUK.stat -> chr.luck++
            }
        }
    }
}